### ISSUE TITLE
WasmFS: Truncate or append doesn't work when writing to a file

### ISSUE THREAD

**Issue by mere-human** [emscripten-core/emscripten#17120]
According to this https://en.cppreference.com/w/c/io/fopen when opening an existing file with `fopen()`, `"w"` should truncate the file and `"a"` should append to it. This doesn't seem  to work when building with WasmFS.

**Version of emscripten/emsdk:**
`emcc (Emscripten gcc/clang-like replacement + linker emulating GNU ld) 3.1.10 (c3fe57af0504fe24fc4ba697feb8c204f3c80022) clang version 15.0.0 (https://github.com/llvm/llvm-project 8bc29d14273b05b05d5a56e34c07948dc2c770d3) Target: wasm32-unknown-emscripten Thread model: posix InstalledDir: ~/src/emsdk/upstream/bin`

**Failing command line in full:**
```cpp
#include <cstdio>
#include <cassert>

int main()
{
    char buffer[100];
    // 1. create
    {
        FILE* fp = fopen("data", "w");
        fputs("123456", fp);
        fclose(fp);
    }
    {
        FILE* fp = fopen("data", "r");
        fgets(buffer, sizeof(buffer), fp);
        fclose(fp);
        printf("%s\n", buffer);  // expect 123456
    }
    // 2. truncate
    {
        FILE* fp = fopen("data", "w");
        fputs("456", fp);
        fclose(fp);
    }
    {
        FILE* fp = fopen("data", "r");
        fgets(buffer, sizeof(buffer), fp);
        fclose(fp);
        printf("%s\n", buffer);  // expect 456
    }
    // 3. append
    {
        FILE* fp = fopen("data", "a");
        fputs("789", fp);
        fclose(fp);
    }
    {
        FILE* fp = fopen("data", "r");
        fgets(buffer, sizeof(buffer), fp);
        fclose(fp);
        printf("%s\n", buffer);  // expect 456789
    }
    return 0;
}
```

**Full link command and output with `-v` appended:**
```bash
emsdk install latest && emsdk activate latest
rm -rf build
mkdir -p build
cd build
em++ -v ../main.cpp -o a.html  -s WASMFS=1
emrun a.html --browser chrome
```
Output with WasmFS:
```
123456
456456
789456
```
Expected output (when building without WasmFS):
```
123456
456
456789
```

**Comment by mere-human** [comment]
The same happens when using `std::fstream`:

```cpp
#include <fstream>
#include <iostream>
#include <string>
#include <cassert>

int main()
{
    std::string s;
    // 1. create
    {
        std::fstream f("data", std::ios::out);
        f << 123456;
    }
    {
        std::fstream f("data");
        f >> s;
        std::cout << s << std::endl; // expect 123456
    }
    // 2. truncate
    {
        std::fstream f("data", std::ios::out | std::ios::trunc);
        f << 456;
    }
    {
        std::fstream f("data");
        f >> s;
        std::cout << s << std::endl; // expect 456
    }
    // 3. append
    {
        std::fstream f("data", std::ios::out | std::ios::app);
        f << 789;
    }
    {
        std::fstream f("data");
        f >> s;
        std::cout << s << std::endl; // expect 456789
    }
    return 0;
}
```

**Comment by mere-human** [comment]
It seems `O_TRUNC` as well as `O_APPEND` are not handled in `doOpen()` here https://github.com/emscripten-core/emscripten/blob/main/system/lib/wasmfs/syscalls.cpp#L377
I'm not sure how this should be handled. Call `doTruncate()`?

**Comment by kripken** [comment]
Thanks for filing, @mere-human !

I think perhaps `doOpen()` can call the underlying API, that is, `.setSize(..)`, since it will have already done the checks beforehand that `doTruncate()` does. Likewise for append it should modify the position directly, I assume. But in both cases maybe when writing the code something will come up.

A PR would be great!

**Comment by tlively** [comment]
This was probably fixed by a combination of #17201 and #17176, but I'll verify that later today.

**Comment by tlively** [comment]
Yep, confirmed that the test program works correctly now.
